pkg/secret/replicated.go

```diff
@@ -0,0 +1,127 @@
+	cleanedClusterName := strings.ReplaceAll(strings.ReplaceAll(clusterName, "_", ""), "-", "")
+	secretName := fmt.Sprintf("%s-superuser", cleanedClusterName)
+
+	return secretName
+}
+
+// VerifySuperuserSecret creates the superUserSecret with propert annotations
+func VerifySuperuserSecret(c client.Client, secretName, namespace string) error {
+	if secretName == "" {
+		return fmt.Errorf("secretName is required")
+	}
+	currentSec := &corev1.Secret{}
+	err := c.Get(context.Background(), types.NamespacedName{Name: secretName, Namespace: namespace}, currentSec)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			password, err := generateRandomString(passwordCharacters, 20)
+			if err != nil {
+				return err
+			}
+
+			sec := &corev1.Secret{
+				ObjectMeta: getManagedObjectMeta(secretName, namespace),
+				Type:       "Opaque",
+				// Immutable feature is only available from 1.21 and up (beta in 1.19 and up)
+				// Immutable:  true,
+				Data: map[string][]byte{
+					"username": []byte(secretName),
+					"password": password,
+				},
+			}
+
+			return c.Create(context.Background(), sec)
+		}
+	}
+
+	// It exists or was created, we won't modify it anymore
+	return nil
+}
+
+// VerifyReplicatedSecret ensures that the correct replicatedSecret for all managed secrets is created
+func VerifyReplicatedSecret(c client.Client, leaderID, namespace string, targetContexts []string) error {
+	// We use leaderID to ensure that this leader instance of k8ssandra-operator is using the correct ReplicatedSecret (it can be shared between different k8ssandra-operators)
+
+	targetRepSec := generateReplicatedSecret(leaderID, namespace, targetContexts)
+	repSec := &api.ReplicatedSecret{}
+	err := c.Get(context.Background(), types.NamespacedName{Name: leaderID, Namespace: namespace}, repSec)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return c.Create(context.Background(), targetRepSec)
+		}
+	}
+
+	// It exists, override whatever was in it
+	currentResourceVersion := repSec.ResourceVersion
+	targetRepSec.DeepCopyInto(repSec)
+	repSec.ResourceVersion = currentResourceVersion
+	return c.Update(context.Background(), repSec)
+}
+
+func generateReplicatedSecret(name, namespace string, targetContexts []string) *api.ReplicatedSecret {
+
+	replicationTargets := make([]api.ReplicationTarget, 0, len(targetContexts))
+	for _, ctx := range targetContexts {
+		replicationTargets = append(replicationTargets, api.ReplicationTarget{
+			K8sContextName: ctx,
+		})
+	}
+
+	return &api.ReplicatedSecret{
+		ObjectMeta: getManagedObjectMeta(name, namespace),
+		Spec: api.ReplicatedSecretSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{api.ManagedByLabel: api.NameLabelValue},
+			},
+			ReplicationTargets: replicationTargets,
+		},
+	}
+}
+
+func getManagedObjectMeta(name, namespace string) metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:      name,
+		Namespace: namespace,
+		Labels: map[string]string{
```

Comment:
I think we should include additional labels, namely `k8ssandra.io/cluster`. I raised the same point for some of the objects created by the stargate controller. Here is the rationale. If every object that the operator creates includes a common label, then a simple label selector query can return an inventory of all objects created on behalf of a K8ssandraCluster. I think this could be useful from a security standpoint.

Reply:
ReplicatedSecret is per k8ssandra-operator deployment, but I'll add that to the superUserSecret.